src/setup_node_env/exit_on_warning.js

```diff
@@ -35,4 +35,16 @@ if (process.noProcessWarnings !== true) {
 
     process.exit(1);
   });
+
+  // While the above warning listener would also be called on
+  // unhandledRejection warnings, we can give a better error message if we
+  // handle them separately:
+  process.on('unhandledRejection', function(reason) {
+    console.error('Unhandled Promise rejection detected:');
+    console.error();
+    console.error(reason);
+    console.error();
+    console.error('Terminating process...');
+    process.exit(1);
+  });
```

Comment:
Isn't this already catched by the `warning` event?

From https://kibana-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/elastic+kibana+pipeline-pull-request/runs/33334/nodes/28/steps/147/log/?start=0

```
Node.js process-warning detected:
[2020-03-16T08:30:26.360Z] 
[2020-03-16T08:30:26.360Z] UnhandledPromiseRejectionWarning: [object Object]
[2020-03-16T08:30:26.360Z]     at emitWarning (internal/process/promises.js:99:15)
[2020-03-16T08:30:26.360Z]     at emitPromiseRejectionWarnings (internal/process/promises.js:143:7)
[2020-03-16T08:30:26.360Z]     at process._tickCallback (internal/process/next_tick.js:69:34)
[2020-03-16T08:30:26.360Z] 
[2020-03-16T08:30:26.360Z] Terminating process...
[2020-03-16T08:30:26.360Z] Fatal error: jest exited with code 1
```

Or will the `reason` catched by the `unhandledRejection` contains more info than what is catched by the `warning` handler?

Reply:
The stack trace available for the unhandled exception is just pointing back to a next-tick frame which doesn't show any of the user-land code. This makes the stack trace unusable and the reader can't see which Promise was not handled correctly.

By implementing an `unhandledRejection` listener we get access to the rejection reason, which hopefully always is an `Error` object with its own stack trace. We can use this stack trace to guide the developer towards the location where the Promise rejection wasn't handled.

Reply:
That was the question :) great then.